server/src/main/java/org/elasticsearch/index/mapper/TimeSeriesIdFieldMapper.java

```diff
@@ -372,8 +379,15 @@ public DocumentDimensions validate(final IndexSettings settings) {
         }
 
         private void add(String fieldName, BytesReference encoded) throws IOException {
-            if (dimensions.add(new Dimension(new BytesRef(fieldName), encoded)) == false) {
-                throw new IllegalArgumentException("Dimension field [" + fieldName + "] cannot be a multi-valued field.");
+            BytesRef name = new BytesRef(fieldName);
+            List<BytesReference> values = dimensions.get(name);
+            if (values == null) {
+                // optimize for the common case where dimensions are not multi-valued
+                values = new ArrayList<>(1);
```

Comment:
This does seem a bit wasteful for the common case where there's only one element. I wonder if it makes sense to keep both the set and the map, and check and use the map only for array values.

Reply:
Not sure if that's worth the complexity. An ArrayList with a single element should't add too much overhead. With this change, I've also removed the `Dimension` wrapper object, so now the list being a wrapper for the value doesn't fundamentally feel different.
One option could be to use `List.of` and replace that with a mutable ArrayList if there's more than one element. That would at least avoid having to allocate another intermediary `Object[]` in the common case. As this method is the only one that modifies the list, I feel like that logic would be well encapsulated.

Reply:
I went and implemented the `List.of` thing.